deepchem/models/normalizing_flows.py

```diff
@@ -0,0 +1,347 @@
+
+    # An instance of tfd.TransformedDistribution
+    self.flow = None
+
+  def __call__(self, *x):
+    return self.flow.bijector.forward(*x)
+
+  @tf.function
+  def fit_on_batch(self, x: np.ndarray,
+                   optimizer: tf.keras.optimizers.Optimizer,
+                   loss: dc.models.losses.Loss) -> float:
+    """Fit on batch of samples.
+    
+    Parameters
+    ----------
+    x: np.ndarray, shape (n_samples, n_dim)
+      Array of samples where each sample is a vector of length `n_dim`.
+    optimizer: dc.models.optimizers.Optimizer
+      An instance of Optimizer.
+    loss: dc.models.losses.Loss
+      An instance of Loss.
+
+    Returns
+    -------
+    batch_loss: float
+      Loss computed on this batch.
+
+    """
+
+    with tf.GradientTape() as tape:
+      dummy_labels = np.ones(len(x))
+      batch_loss = loss(x, dummy_labels)
+      grads = tape.gradient(batch_loss, self.trainable_variables)
+      optimizer.apply_gradients(zip(grads, self.trainable_variables))
+    return batch_loss
+
+
+class NormalizingFlowModel(NormalizingFlow):
+  """A base distribution and normalizing flow for applying transformations.
+
+  A distribution implements two main operations:
+    1. Sampling from the transformed distribution.
+    2. Calculating log probabilities.
+
+  A normalizing flow implements three main operations:
+    1. Forward transformation, 2. Inverse transformation, and 
+    3. Calculating the Jacobian.
+
+  Deep Normalizing Flow models require normalizing flow layers where
+  input and output dimensions are the same, the transformation is invertible,
+  and the determinant of the Jacobian is efficient to compute and
+  differentiable. The determinant of the Jacobian of the transformation 
+  gives a factor that preserves the probability volume to 1 when transforming
+  between probability densities of different random variables.
+
+  They are effective for any application requiring a probabilistic
+  model with these capabilities, e.g. generative modeling,
+  unsupervised learning, or probabilistic inference. For a thorough review
+  of normalizing flows, see [1]_.
+
+  References
+  ----------
+  .. [1] Papamakarios, George et al. "Normalizing Flows for Probabilistic Modeling and Inference." (2019). https://arxiv.org/abs/1912.02762.
+
+  """
+
+  def __init__(self,
+               base_distribution,
+               flow_layers: Sequence,
+               optimizer: Optional[tf.keras.optimizers.Optimizer] = None,
+               loss: Optional[Any] = None,
+               **kwargs):
+    """Creates a new NormalizingFlowModel.
+
+    Parameters
+    ----------
+    base_distribution : tfd.Distribution
+      Probability distribution to be transformed.
+      Typically an N dimensional multivariate Gaussian.
+    flow_layers : Sequence[tfb.Bijector]
+      An iterable of bijectors that comprise the flow.
+    optimizer: Optional[tf.keras.optimizers.Optimizer]
+      An instance of Optimizer.
```

Comment:
Can we make it accept `dc.models.Optimizar` objects?  We should also have a `learning_rate` argument.